cumulusci/tasks/connectedapp.py

```diff
@@ -0,0 +1,174 @@
+    </types>
+    <version>44.0</version>
+</Package>"""
+
+
+class CreateConnectedApp(SFDXBaseTask):
+    task_options = {
+        "label": {
+            "description": "The label for the connected app.  Must contain only alphanumeric and underscores",
+            "required": True,
+        },
+        "email": {
+            "description": "The email address to associate with the connected app.  Defaults to email address from the github service if configured.",
+            "required": True,
+        },
+        "username": {
+            "description": "Create the connected app in a different org.  Defaults to the defaultdevhubusername configured in sfdx.",
+            "required": False,
+        },
+        "connect": {
+            "description": "If True, the created connected app will be stored as the CumulusCI connected_app service in the keychain.",
+            "required": False,
+        },
+        "overwrite": {
+            "description": "If True, any existing connected_app service in the CumulusCI keychain will be overwritten.  Has no effect if the connect option is False.",
+            "required": False,
+        },
+    }
+
+    def _init_options(self, kwargs):
+        self.client_id = None
+        self.client_secret = None
+        kwargs["command"] = "force:mdapi:deploy --wait 5"
+        super(CreateConnectedApp, self)._init_options(kwargs)
+
+        # Validate label
+        if not re.match(r"^\w+$", self.options["label"]):
+            raise TaskOptionsError(
+                "label value must contain only alphanumeric or underscore characters"
+            )
+
+        # Default email to the github service's email if configured
+        if "email" not in self.options:
+            try:
+                github = self.project_config.keychain.get_service("github")
+            except ServiceNotConfigured:
+                raise TaskOptionsError(
+                    "Could not determine a default for option 'email'.  Either configure the github service using 'cci service connect github' or provide a value for the 'email' option"
+                )
+            self.options["email"] = github.email
+
+        # Default to sfdx defaultdevhubusername
+        if "username" not in self.options:
+            self._set_default_username()
+        self.options["command"] += " -u {}".format(self.options.get("username"))
+
+        self.options["connect"] = process_bool_arg(self.options.get("connect", False))
+        self.options["overwrite"] = process_bool_arg(
+            self.options.get("overwrite", False)
+        )
+
+    def _set_default_username(self):
+        self.logger.info("Getting username for the default devhub from sfdx")
+        self._run_command(
+            command="{} force:config:get defaultdevhubusername --json".format(SFDX_CLI),
+            env=self._get_env(),
+            output_handler=self._process_devhub_output,
+        )
+
+    def _process_json_output(self, line):
```

Comment:
this could be static, or even just a function in utils

Reply:
I guess it could but then you'd have to pass it the instance with the logger to emit a log line before raising the exception.  It seems like this is a particular use case.  There's nothing special it's doing other than simply decoding json and deciding how to handle an exception.